In [1]:
# Let's load in the data we want
import pandas as pd

fr_train = pd.read_csv('../input/train.csv')
print(fr_train.shape)
fr_train.head()

(891, 12)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

So right off the bat we can see quite a few missing values in the 'Cabin' variable. It's always good to find and clean up missing or clearly incorrect values when you're working with a dataset, so let's do that first.

In [2]:
def naSummary(df):
    nrow, ncol = df.shape
    na_count = df.isnull().sum()
    na_pc = na_count.divide(nrow)
    print(pd.DataFrame({'NA Count': na_count, 'NA %': na_pc}))
    
naSummary(fr_train)

                 NA %  NA Count
PassengerId  0.000000         0
Survived     0.000000         0
Pclass       0.000000         0
Name         0.000000         0
Sex          0.000000         0
Age          0.198653       177
SibSp        0.000000         0
Parch        0.000000         0
Ticket       0.000000         0
Fare         0.000000         0
Cabin        0.771044       687
Embarked     0.002245         2


Looks like we're missing data in three columns: Age, Cabin and Embarked. Fortunately we're not missing too much Age data and practically none in Embarked. On the other hand we're missing a whopping 687 values for Cabin, or as we can see from the 'NA %' variable, we're missing Cabin values for 77% of our entire dataset!

Since we're only missing a measly two values for 'Embarked' lets try and rectify those. First let's find the relevant passengers.

In [3]:
# Get the indices of the rows with the missing 'Embarked' values
fr_train[fr_train.isnull()['Embarked']]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

Looks like they share a ticket number, my original suspicion was that this was erroneous data but turns out the Titanic offerd group type tickets. Typically this points out families as we can see in the simple example below.

In [4]:
fr_train[fr_train.duplicated(subset='Ticket')]
fr_train[fr_train['Ticket'] == '349909']

PassengerId  Survived  Pclass  \
7              8         0       3   
24            25         0       3   
374          375         0       3   
567          568         0       3   

                                            Name     Sex   Age  SibSp  Parch  \
7                 Palsson, Master. Gosta Leonard    male   2.0      3      1   
24                 Palsson, Miss. Torborg Danira  female   8.0      3      1   
374                   Palsson, Miss. Stina Viola  female   3.0      3      1   
567  Palsson, Mrs. Nils (Alma Cornelia Berglund)  female  29.0      0      4   

     Ticket    Fare Cabin Embarked  
7    349909  21.075   NaN        S  
24   349909  21.075   NaN        S  
374  349909  21.075   NaN        S  
567  349909  21.075   NaN        S

So if we're lucky there might be other people on the same ticket with the Embarkment value already set, and seeing as they shared a ticket number I think it would be safe to assume they embarked from the same port. I don't really expect this to be the case but it's a simple check so why not!

In [5]:
fr_train[fr_train['Ticket'] == '113572']

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

Turns out we're not so lucky.

So we still need to assign them an Embarkment value. Now you could do this in a more statistical manner, formally called 'unit imputation', but why overcomplicate things? They both survived and there's a plethora of information on Titanic survivors available. A quick google search of their respective names yields the following information on courtesy of Encyclopedia Titanica.

> Mrs Stone boarded the Titanic in Southampton on 10 April 1912 and was travelling in first class with her maid Amelie Icard.

Easy!

In [6]:
fr_train.set_value([61, 829], 'Embarked', 'S');

# Verify by using our naSummary
naSummary(fr_train)

                 NA %  NA Count
PassengerId  0.000000         0
Survived     0.000000         0
Pclass       0.000000         0
Name         0.000000         0
Sex          0.000000         0
Age          0.198653       177
SibSp        0.000000         0
Parch        0.000000         0
Ticket       0.000000         0
Fare         0.000000         0
Cabin        0.771044       687
Embarked     0.000000         0


So there go our missing Embarked values. So in the interest of common sense we should realise that when we only have 2 missing values out of 891 we probably could have just done anything we want with them and statistically it wouldn't have made a difference. Mean value imputation would have been more than suitable for this (replace the missing values with the overall mean for that vector), and it actually would've made given us the same result in this case.

Regardless let's move onto filling the missing age values.

* Write about my script magics here*

In [7]:
fr_train_plus = pd.read_csv('../input/train_plus.csv')
print(fr_train.shape)
fr_train.head()

(891, 12)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [8]:
naSummary(fr_train_plus)

                 NA %  NA Count
PassengerId  0.000000         0
Survived     0.000000         0
Pclass       0.000000         0
Name         0.000000         0
Sex          0.000000         0
Age          0.000000         0
SibSp        0.000000         0
Parch        0.000000         0
Ticket       0.000000         0
Fare         0.000000         0
Cabin        0.771044       687
Embarked     0.002245         2


Excellent! No more missing Age values, let's double check that there's no bug in our scripts which overwrote original Age data quickly.

In [9]:
checkv = fr_train['Age'] != fr_train_plus['Age']
ogv = fr_train['Age'].isnull()

(checkv == ogv).all()

True

Awesome! The "checkv" vector is a series of bools which are True when there's a difference in the original data and my new data, the "ogv" vector is a series of bools which are True when there's a missing value in the original data. This code confirms that both vectors are identical, meaning my script only replaced missing data and all the original data is intact.

Though look and remember that we've reintroduced our missing Embarked values, we'll have to fix that again!

In [10]:
fr_train_plus.set_value([61, 829], 'Embarked', 'S');
naSummary(fr_train_plus)

                 NA %  NA Count
PassengerId  0.000000         0
Survived     0.000000         0
Pclass       0.000000         0
Name         0.000000         0
Sex          0.000000         0
Age          0.000000         0
SibSp        0.000000         0
Parch        0.000000         0
Ticket       0.000000         0
Fare         0.000000         0
Cabin        0.771044       687
Embarked     0.000000         0


Now all we have to deal with is our missing Cabin values. 

In [14]:
fr_train_plus[fr_train_plus['Cabin'].notnull()]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
21            22         1       2   
23            24         1       1   
27            28         0       1   
31            32         1       1   
52            53         1       1   
54            55         0       1   
55            56         1       1   
61            62         1       1   
62            63         0       1   
66            67         1       2   
75            76         0       3   
88            89         1       1   
92            93         0       1   
96            97         0       1   
97            98         1       1   
102          103         0       1   
110          111         0       1   
118          119         0       1   
123          124         1       2   
124          125         0       1   
128          129         1       3   
136          137         1       1   
137          138         0       1   
139          140         0       1   
148          149         0       2   
..           ...       ...     ...   
751          752         1       3   
759          760         1       1   
763          764         1       1   
765          766         1       1   
772          773         0       2   
776          777         0       3   
779          780         1       1   
781          782         1       1   
782          783         0       1   
789          790         0       1   
796          797         1       1   
802          803         1       1   
806          807         0       1   
809          810         1       1   
815          816         0       1   
820          821         1       1   
823          824         1       3   
829          830         1       1   
835          836         1       1   
839          840         1       1   
849          850         1       1   
853          854         1       1   
857          858         1       1   
862          863         1       1   
867          868         0       1   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
21                               Beesley, Mr. Lawrence    male  34.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
27                      Fortune, Mr. Charles Alexander    male  19.0      3   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female  57.0      1   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
54                      Ostby, Mr. Engelhart Cornelius    male  65.0      0   
55                                   Woolner, Mr. Hugh    male  46.0      0   
61                                 Icard, Miss. Amelie  female  38.0      0   
62                         Harris, Mr. Henry Birkhardt    male  45.0      1   
66                        Nye, Mrs. (Elizabeth Ramell)  female  29.0      0   
75                             Moen, Mr. Sigurd Hansen    male  25.0      0   
88                          Fortune, Miss. Mabel Helen  female  23.0      3   
92                         Chaffee, Mr. Herbert Fuller    male  46.0      1   
96                           Goldschmidt, Mr. George B    male  71.0      0   
97                     Greenfield, Mr. William Bertram    male  23.0

In [ ]:
fr_trin